exercises/05-reselect/src/modules/users/components/user-list.tsx

```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+
+import { useSelector, useDispatch } from 'react-redux'
+
+import { usersActions } from 'modules/users/users-slice'
+import { UserData } from '../user-types'
+import { getUserList } from '../users-selectors'
+
+export const UserList: React.FC = () => {
+  const users = useSelector(getUserList)
+  const dispatch = useDispatch()
+  const addUser = (user: UserData) => () => dispatch(usersActions.addUser(user))
```

Comment:
I'd probably go with `useCallback` somewhere in the middle - kinda harder with this high order function approach. But keeping references constant is really a good habit. 

Also i believe that `addUser` is not good naming as it doesn't technically add the user - only creates the handler, that needs to be called to add user.

Reply:
Is there a benefit in using `useCallback` in this situation? The documentation says that it's main purpose is to create a memoized callback, which is to be passed to children components in order to avoid unnecessary rerendering. This is not the case here and at first glance it seems unnecessary to me. Moreover, it seems less readable to me:
```ts
const addAryaCallback = useCallback(
    () =>
      dispatch(usersActions.addUser({ firstName: 'Arya', lastName: 'Stark' })),
    [dispatch]
  )
// ...
<button onClick={addAryaCallback}>Add No One</button>
```

How about this approach? Would you approve this?
```tsx
const dispatchAddUser = (user: UserData) => dispatch(usersActions.addUser(user))
// ...
<button
    onClick={() =>
        dispatchAddUser({ firstName: 'Daenerys', lastName: 'Targaryen' })
    }
>
     Add Mother of Dragons
</button>
```

Reply:
I've reorganized the code according to my suggestion, see 14bf4f1

Reply:
The problem typically occurs in later stages of development. Now it's just a button, later someone can turn this into component without stable reference. That's why it's always good practice to keep everything memoized, since once a single reference is not, it may cause a huge perf implications later on.

Reply:
But at this stage of the development this is just a button and the memoization adds no value, while additional code (`useCallback` logic) must be executed every time. The memoization may be added anytime when and **if** it's needed (e.g. a button shall be replaced with a separate component). I don't think it's needed here.

Reply:
OK, I see it now. I've seen somewhere that in such simple case the usage of `useCallback` might be worse from the performance point of view and got it confused.

I'll add it tomorrow. But, since this piece of code comes from exercise 4, I think it might deserve a separate PR.

Reply:
I'm fine with both approaches here. Actually, I wanted to suggest `useCallback` in the previous PRs but I was waiting for a better, more complicated component. Anyway, since this is a training project, we should introduce `useCallback` in the correct exercise that introduces it in the presentation.

Reply:
Thank you for the remark!

Done in 5cececc.

I've decided to add this to exercise 5 and latter and to not modify previous exercises. The reason is that the memoization isn't mentioned in slides up util exercise 5 - reselect. I think `useCallback` will fit well to this chapter. 

I'm working on slides now.

Reply:
I have some questions.
I've been investigating some more while trying to come up with a meaningful example for slides. 

1. It seems at the first glance, that if I'd like to use memoization for own components, they must be created with `React.memo`, is it correct?
2. If yes, shall it be mentioned in the slides? 
3. When I pass a callback created using `useCallback` to a simple button, will it be re-rendered with the parent component or not? Is the `<button>` memoized? Is there a way to know?

Reply:
I've added page [50](https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs/edit#slide=id.g9003da311f_0_7) to slides. It's not very informative. Feel free to edit it in any way you see fit

Reply:
1. Yes.
2. It would be great to mention it. :+1: 
3. I'm not sure how components for DOM elements work but I trust you that they are re-rendered every time. :slightly_smiling_face: It makes sense because it's default behavior.

Reply:
2. I'll add it to slides if needed (see below)
3. Does it mean, that the `useCallback` is kinda useless here with an ordinary `<button>`? When I try the same thing with a custom component created with `React.memo` it's not re-rendered.

In such case what shall I do with the tutorial?
1. `useCallback` + `<button>`
2. `useCallback` + custom component which wraps button
3. no `useCallback` + `<button>`

Reply:
Yes, it seems to be useless in this particular use case with a plain `<button>` based on your experiments. However, in most real apps, almost everyone uses custom components for buttons and everything so it's beneficial to use it in general.

> In such case what shall I do with the tutorial?

It would be great to do the 2nd option so trainees can see the benefit.

IIRC, we have some optional/extra task for `Component` vs `PureComponent`. Might be good to combine this together or at least include `React.memo` into the extra task. We could use `useCallback` here with a note that it's useful only with `React.memo` or `PureComponent`.

Reply:
I've rewritten the tutorial and solution with `useCallback` and a memoized version of button, see 06d6bec. I hope it's OK.

I'll update slides as soon as possible.

I believe it's the last remark for this PR.

Reply:
- Added slide [React.memo](https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs/edit#slide=id.g9003da311f_0_7)
- Modified slide [`useCallback`](https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs/edit#slide=id.g9041b3b803_0_2)

Reply:
@novellizator  as agreed, fixed typing https://github.com/salsita/react-training/pull/53/commits/ddb02e271ecee7c2ea4259a5b5cd499299c3b902 and README typo https://github.com/salsita/react-training/pull/53/commits/c6659990d060f148fa24834b821859dc2f793862. Will merge today